python/lsst/cp/pipe/cpSpectroFlat.py

```diff
@@ -0,0 +1,324 @@
+        inputs = butlerQC.get(inputRefs)
+
+        # Get provenance.  This is essential for debugging.
+        inputMD = inputs['inputFlat'].getMetadata()
+        for inputName in sorted(list(inputs.keys())):
+            reference = getattr(inputRefs, inputName, None)
+            if reference is not None and hasattr(reference, "run"):
+                runKey = f"LSST CALIB RUN {inputName.upper()}"
+                runValue = reference.run
+                idKey = f"LSST CALIB UUID {inputName.upper()}"
+                idValue = str(reference.id)
+
+                inputMD[runKey] = runValue
+                inputMD[idKey] = idValue
+        inputs['inputMD'] = inputMD
+
+        outputs = self.run(**inputs)
+        butlerQC.put(outputs, outputRefs)
+
+    def run(self, dummyExpRef, inputFlat, inputPtc, inputMD):
+        """Create spectro flat from an input flat and PTC.
+
+        Parameters
+        ----------
+        dummyExpRef : `lsst.daf.butler.DeferredDatasetHandle`
+            Exposure used to ensure the correct flat is chosen.
+        inputFlat : `lsst.afw.image.Exposure`
+            Input flat to convert to spectro-flat.
+        inputPtc : `lsst.ip.isr.PhotoTransferCurveDataset`
+            Input PTC containing the gains to apply to spectro-flat.
+        inputMD : `lsst.daf.base.PropertyList`
+            Provenance metadata derived from the butler references.
+
+        Returns
+        -------
+        results : `lsst.pipe.base.Struct`
+            The results struct containing:
+
+            ``outputData``
+                Final modified flat generated from the inputs
+                (`lsst.afw.image.Exposure`).
+        """
+        # Verify input
+        exposureFilter = dummyExpRef.get(component="filter")
+        if exposureFilter.physicalLabel != self.config.inputFlatPhysicalFilter:
+            raise RuntimeError("Did not receive expected physical_filter: "
+                               f"{exposureFilter.physicalLabel} != "
+                               f"{self.config.inputFlatPhysicalFilter}")
+
+        detector = inputFlat.getDetector()
+        statsControl = afwMath.StatisticsControl(self.config.backgroundSigmaClipThreshold,
+                                                 5,
+                                                 0x0)
+
+        # Make gain flat
+        # https://github.com/lsst/atmospec/blob/u/jneveu/doFlat/python/lsst/atmospec/spectroFlat.py#L85
+        # TODO: DM-46080 Ensure cpSpectroFlat can distinguish between
+        # new and old flats.
+        gains = inputPtc.gain
+        scaledGains = {}
+        gainFlat = inputFlat.clone()
+
+        # Todo: Get temperature dependence terms.
+        if self.config.scaleGainsByMean:
+            gainMean = np.mean(list(gains.values()))
+        else:
+            gainMean = 1.0
+
+        for amp in detector:
+            bbox = amp.getBBox()
+            gainFlat[bbox].image.array[:, :] = 1 / (gains[amp.getName()] / gainMean)
```

Comment:
Isn't gainFlat = scaledGains rather than 1/scaledGains? It is just an exposure with the gain so it should just be the gain rescaled.

Reply:
Not based on the example code/notebook supplied.  The original method ( https://github.com/lsst/atmospec/blob/u/jneveu/doFlat/python/lsst/atmospec/spectroFlat.py#L120 ) has an `invertGains` option that allows the gains to be in the denominator (if `True`), or in the numerator (if `False`).  The defaults have this set to `False`, but the example notebook has this flipped.  

I see now that I added the config option `applyGainDirection`, but never used it.  I've fixed that, correcting the docstring to indicate which direction `True` denotes, and wrapped this operation in an if/else block.

I think this complicates the next ticket, that will update this for the task change, but I haven't thought it completely through yet.  I did add some more details on the JIRA ticket, to preserve my thinking from today.